subprojects/core/src/main/java/org/gradle/caching/internal/DefaultBuildCacheKeyBuilder.java

```diff
@@ -135,69 +124,12 @@ private static void log(String type, Object value) {
     }
 
     public BuildCacheKeyBuilder appendToCacheKey(Object value) {
-
-        if (value == null) {
-            this.putString("$NULL");
-            return this;
-        }
-
-        if (value.getClass().isArray()) {
-            this.putString("Array");
-            for (int idx = 0, len = Array.getLength(value); idx < len; idx++) {
-                this.putInt(idx);
-                this.appendToCacheKey(Array.get(value, idx));
-            }
-            return this;
+        try {
+            HasherUtil.putObject(hasher, value);
+        } catch (NotSerializableException e) {
+            throw new UncheckedIOException(e);
```

Comment:
Could you elaborate on this, I'm not sure I fully understand?

The previous code would simply try to cast all `Object` to `Serializable` and use `SerializationUtils.serialize` as a last resort action. However, if the class isn't `Serializable` it will throw a `ClassCastException`. I simply tried to mimic the behavior from how input property was previously serialized without changing the behavior for `BuildCacheKey`. If the class isn't `Serializable` it throws a `NotSerializableException` so we can handle it as needed for `BuildCacheKey` and `InputProperty`.